diagnostics/diagnostic.go

```diff
@@ -468,6 +474,59 @@ func (d *DiagnosticClient) runSnapshotFilesListListener() {
 	}()
 }
 
+func appendWithCap(list *list.List, cap int, items []time.Duration) {
+	if list.Len() == cap {
+		list.Remove(list.Front())
+	}
+
+	for _, v := range items {
+		list.PushBack(v)
+	}
+}
+
+func (d *DiagnosticClient) runBlockMetricsListener() {
+	go func() {
+		ctx, ch, cancel := diaglib.Context[diaglib.AppendBlockMetrics](context.Background(), 1)
+		defer cancel()
+
+		rootCtx, _ := common.RootContext()
+
+		diaglib.StartProviders(ctx, diaglib.TypeOf(diaglib.AppendBlockMetrics{}), log.Root())
+		for {
+			select {
+			case <-rootCtx.Done():
+				cancel()
+				return
+			case info := <-ch:
+				d.mu.Lock()
+
+				if info.Header != nil {
```

Comment:
when working with slices prefer to use `len` instead of nil checks (len(nil) is 0), for example
```
if len(info.Header) > 0 { do stuff } 
```